packages/langium/src/serializer/json-serializer.ts

```diff
@@ -130,6 +141,12 @@ export class DefaultJsonSerializer implements JsonSerializer {
                 astNode ??= { ...value };
                 astNode.$sourceText = value.$cstNode?.text;
             }
+            if (comments && isAstNode(value)) {
+                astNode ??= {
+                    ...value,
+                    $comment: this.commentProvider.getComment(value)
+                } as AstNodeWithComment;
```

Comment:
This will fail in case either `textRegions` or `sourceText` is true. The right side of `??=` is only evaluated when the left side is `undefined`. Therefore, if it's not undefined, `$comment` will not be set.

Reply:
Something is wrong with this snippet:
```ts
else {
  let astNode: AstNodeWithTextRegion | undefined = undefined;
  if (textRegions && isAstNode(value)) {
      astNode = this.addAstNodeRegionWithAssignmentsTo({ ...value });
      if ((!key || value.$document) && astNode?.$textRegion) {
          try {
              astNode.$textRegion.documentURI = getDocument(value).uri.toString();
          } catch (e) { /* do nothing */ }
      }
  }
  if (sourceText && !key && isAstNode(value)) {
      astNode ??= { ...value };
      astNode.$sourceText = value.$cstNode?.text;
  }
  if (comments && isAstNode(value)) {
      astNode ??= {
          ...value,
          $comment: this.commentProvider.getComment(value)
      } as AstNodeWithComment;
  }
  return astNode ?? value;
}
```

When `textRegions` and `sourceText` are true, only `astNode.$sourceText` will be set. And I did some similar mistake by copy-pasting.

Reply:
Ah! No sorry, now I see it. Very weird way of writing this